# FinMark Corporation - Data Cleaning & EDA

## 📌 Project Overview
FinMark Corporation is a financial services company that wants to strengthen its ability to generate insights from customer data.

As the data analyst, your role is to build the foundation for a **machine learning solution** by first ensuring data quality and conducting exploratory data analysis (EDA).

This notebook focuses on **Milestone 1** of the project: cleaning the data and performing an initial EDA.

The work here will prepare the datasets for visualization and model development in later stages.


### Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
plt.style.use("ggplot")
sns.set_palette("Set2")

In [12]:
demographics = pd.read_csv("CLEANED_customer_demographics.csv")
transactions = pd.read_csv("CLEANED_customer_transactions.csv")
social = pd.read_csv("clean_social_media_interactions.csv")

In [4]:
demographics

,CustomerID,Age,Gender,Location,IncomeLevel,SignupDate
0,9207fa75-5758-48d1-94ad-19c041e0520f,51.0,Female,Jensenberg,Low,2022-11-17
1,5fb09cd8-a473-46f7-80bd-6e49cf509078,NaN,Female,Castilloport,High,2020-07-21
2,c139496e-cc89-498a-bd90-1fb4627b6cff,37.0,Male,Lake Jennifertown,NaN,2021-01-01
3,50118139-7264-428f-81cc-a25fddc5d6dd,44.0,Male,Port Carl,Medium,2024-06-10
4,7d1f2bbc-8d16-4fbc-9b37-ece3324e8ed4,50.0,Female,Jessebury,High,2023-08-24
...,...,...,...,...,...,...
3018,48313b18-ccdb-4946-b958-1ebfc92dcc2a,59.0,Male,South Andrea,High,2020-02-26
3019,bbe76a26-234f-4a70-b71f-fa9b183f6353,Unknown,Male,Gregoryfurt,High,2020-06-23
3020,7eb7f7df-9b0a-416a-829e-bcf4d8bd57dc,55.0,Male,New Anthonyland,High,2022-07-23
3021,ce5cf079-859b-43fa-9609-c6bb7e12c6b7,41.0,Female,Adrianhaven,Low,2024-01-10


In [5]:
transactions

,CustomerID,TransactionID,TransactionDate,Amount,ProductCategory,PaymentMethod
0,60567026-f719-4cd6-849e-137e86d8938f,5ff75116-0a50-4d04-80fb-31e5ccbb0769,2024-05-15,117.64,Clothing,PayPal
1,4090ba85-b111-4f75-a792-c777965f5255,2c39b9fe-ff57-4d39-9321-9f5cdf187aa1,2023-04-26,466.14,Health & Beauty,Bank Transfer
2,9223891b-73ff-4d5c-b8ae-13ece82ee28b,f79588dd-3db9-4ffa-97f8-7de0e64259f1,2022-09-23,563.99,Clothing,Debit Card
3,9243eebc-938f-480c-8564-16d503d250de,401c0fc9-60df-4455-ad78-67c132f9897d,2024-04-15,254.44,Automotive,PayPal
4,6e3e8eb8-bc0f-4ffe-9f74-5d5efec9502f,2034aebc-8280-4254-a667-92bcd1c2be4f,2024-06-03,590.52,Home & Garden,Bank Transfer
...,...,...,...,...,...,...
3010,086370e7-8884-47d7-89b0-776401a43e1c,652636b6-5c72-44e1-aefb-6adce412ee75,2023-12-05,974.25,Electronics,Debit Card
3011,7c7df909-aa25-48c7-af9d-6ebe714b28b6,eb26a4f4-5af2-4e25-b1f2-fd0de5583f7b,10/08/2022,50.07,Automotive,Bank Transfer
3012,79fc606a-c02c-4c39-b065-3e88a59b366c,7fbe5cc4-a35e-4e8e-89d7-7b439659b9f6,20/07/2022,272.68,Clothing,Debit Card
3013,6a84f1a7-f636-4300-bd8d-c7f888460829,185ee140-348a-4461-bd40-07a7e8535815,2023-10-09,0,Health & Beauty,PayPal


In [6]:
social

,CustomerID,InteractionID,InteractionDate,Platform,InteractionType,Sentiment
0,2dcb9523-356b-40b2-a67b-1f27797de261,e5d15761-d0a7-4329-89e3-79a892c56097,2023-07-11,NaN,Comment,NaN
1,e12c37b3-7d4d-472f-9fd8-0df2cb3001aa,02f9f376-70ae-4fcd-9070-1db977939948,2023-07-06,Twitter,Share,NaN
2,08a911a3-65e6-4f5d-a6a1-ae7ddcbe28a2,a83fa04c-f109-4f24-8ce1-2078154f6a1c,2024-05-24,Instagram,Comment,Neutral
3,efdfdfc9-5dbb-4478-911a-101a390a0285,28a69c4b-a2e4-4c74-a130-1132d7733fdf,2023-11-01,Instagram,Like,Neutral
4,ca1e90f6-0e5f-492e-ab92-252ff540da18,d9d1c6f8-5e15-4738-b52b-13c2982420cc,2023-07-08,Instagram,Like,NaN
...,...,...,...,...,...,...
3015,e06f0e16-8e48-483b-b103-9f105b139ca8,071a2657-12a8-4216-9c64-baf9ab5c9420,2024-04-30,Twitter,Share,Neutral
3016,4d20c2b1-2f8a-4066-902c-bd068c2adccc,e5be1d7a-fdba-4e91-8674-577ac01cddf0,2024-05-16,NaN,Like,Very Negative
3017,b7b84b81-c61b-4f23-b3db-b18010484122,93d2d3e1-841e-4ece-91c1-3b1998a4cab8,2023-12-16,Twitter,Like,Neutral
3018,5659a8c6-0397-4396-b561-52ec1c8a760b,d7f6806b-f348-431e-9198-07f0685d8d8e,2024-01-28,Facebook,Share,Negative


## Data Cleaning

In [13]:
print("Null values per dataset:")
print("Demographics:\n", demographics.isnull().sum())
print("Transactions:\n", transactions.isnull().sum())
print("Social Media:\n", social.isnull().sum())

Null values per dataset:
Demographics:
 CustomerID       0
Age            276
Gender           0
Location         0
IncomeLevel    285
SignupDate       0
dtype: int64
Transactions:
 CustomerID           0
TransactionID        0
TransactionDate      0
Amount             283
ProductCategory    282
PaymentMethod        0
dtype: int64
Social Media:
 CustomerID           0
InteractionID        0
InteractionDate      0
Platform           291
InteractionType      0
Sentiment          309
dtype: int64


In [14]:
# Convert Age to numeric before imputation
demographics["Age"] = pd.to_numeric(demographics["Age"], errors="coerce")
demographics["Age"] = demographics["Age"].fillna(demographics["Age"].median())

In [15]:
# Convert Amount to numeric, coerce errors if any
transactions["Amount"] = pd.to_numeric(transactions["Amount"], errors="coerce")
transactions["Amount"] = transactions["Amount"].fillna(transactions["Amount"].median())

In [16]:
# Fill missing values in demographics
demographics["Age"] = demographics["Age"].fillna(demographics["Age"].median())
demographics["IncomeLevel"] = demographics["IncomeLevel"].fillna("Unknown")

# Fill missing values in transactions
transactions["Amount"] = transactions["Amount"].fillna(transactions["Amount"].median())
transactions["ProductCategory"] = transactions["ProductCategory"].fillna("Unknown")

# Fill missing values in social media
social["Platform"] = social["Platform"].fillna("Unknown")
social["Sentiment"] = social["Sentiment"].fillna("Unknown")

# Double-check if nulls remain
print("Remaining nulls after imputation:")
print("Demographics:\n", demographics.isnull().sum())
print("Transactions:\n", transactions.isnull().sum())
print("Social Media:\n", social.isnull().sum())

Remaining nulls after imputation:
Demographics:
 CustomerID     0
Age            0
Gender         0
Location       0
IncomeLevel    0
SignupDate     0
dtype: int64
Transactions:
 CustomerID         0
TransactionID      0
TransactionDate    0
Amount             0
ProductCategory    0
PaymentMethod      0
dtype: int64
Social Media:
 CustomerID         0
InteractionID      0
InteractionDate    0
Platform           0
InteractionType    0
Sentiment          0
dtype: int64


In [18]:
# Remove unrealistic ages
demographics.loc[(demographics["Age"] < 18) | (demographics["Age"] > 100), "Age"] = np.nan

# Fill again with median
demographics["Age"] = demographics["Age"].fillna(demographics["Age"].median())

# Check after cleaning
print(demographics["Age"].describe())

count    3023.000000
mean       44.688720
std        14.424064
min        18.000000
25%        34.000000
50%        45.000000
75%        57.000000
max        70.000000
Name: Age, dtype: float64


#### Save the cleaned file

In [19]:
demographics.to_csv("CLEANED_customer_demographics_v2.csv", index=False)
transactions.to_csv("CLEANED_customer_transactions_v2.csv", index=False)
social.to_csv("clean_social_media_interactions_v2.csv", index=False)

print("Cleaned datasets saved as *_v2.csv")

Cleaned datasets saved as *_v2.csv


In [20]:
demographics_v2 = pd.read_csv("/content/CLEANED_customer_demographics_v2.csv")
transactions_v2 = pd.read_csv("/content/CLEANED_customer_transactions_v2.csv")
social_v2 = pd.read_csv("/content/clean_social_media_interactions_v2.csv")

In [21]:
demographics_v2

,CustomerID,Age,Gender,Location,IncomeLevel,SignupDate
0,9207fa75-5758-48d1-94ad-19c041e0520f,51.0,Female,Jensenberg,Low,2022-11-17
1,5fb09cd8-a473-46f7-80bd-6e49cf509078,45.0,Female,Castilloport,High,2020-07-21
2,c139496e-cc89-498a-bd90-1fb4627b6cff,37.0,Male,Lake Jennifertown,Unknown,2021-01-01
3,50118139-7264-428f-81cc-a25fddc5d6dd,44.0,Male,Port Carl,Medium,2024-06-10
4,7d1f2bbc-8d16-4fbc-9b37-ece3324e8ed4,50.0,Female,Jessebury,High,2023-08-24
...,...,...,...,...,...,...
3018,48313b18-ccdb-4946-b958-1ebfc92dcc2a,59.0,Male,South Andrea,High,2020-02-26
3019,bbe76a26-234f-4a70-b71f-fa9b183f6353,45.0,Male,Gregoryfurt,High,2020-06-23
3020,7eb7f7df-9b0a-416a-829e-bcf4d8bd57dc,55.0,Male,New Anthonyland,High,2022-07-23
3021,ce5cf079-859b-43fa-9609-c6bb7e12c6b7,41.0,Female,Adrianhaven,Low,2024-01-10


In [22]:
transactions_v2

,CustomerID,TransactionID,TransactionDate,Amount,ProductCategory,PaymentMethod
0,60567026-f719-4cd6-849e-137e86d8938f,5ff75116-0a50-4d04-80fb-31e5ccbb0769,2024-05-15,117.64,Clothing,PayPal
1,4090ba85-b111-4f75-a792-c777965f5255,2c39b9fe-ff57-4d39-9321-9f5cdf187aa1,2023-04-26,466.14,Health & Beauty,Bank Transfer
2,9223891b-73ff-4d5c-b8ae-13ece82ee28b,f79588dd-3db9-4ffa-97f8-7de0e64259f1,2022-09-23,563.99,Clothing,Debit Card
3,9243eebc-938f-480c-8564-16d503d250de,401c0fc9-60df-4455-ad78-67c132f9897d,2024-04-15,254.44,Automotive,PayPal
4,6e3e8eb8-bc0f-4ffe-9f74-5d5efec9502f,2034aebc-8280-4254-a667-92bcd1c2be4f,2024-06-03,590.52,Home & Garden,Bank Transfer
...,...,...,...,...,...,...
3010,086370e7-8884-47d7-89b0-776401a43e1c,652636b6-5c72-44e1-aefb-6adce412ee75,2023-12-05,974.25,Electronics,Debit Card
3011,7c7df909-aa25-48c7-af9d-6ebe714b28b6,eb26a4f4-5af2-4e25-b1f2-fd0de5583f7b,10/08/2022,50.07,Automotive,Bank Transfer
3012,79fc606a-c02c-4c39-b065-3e88a59b366c,7fbe5cc4-a35e-4e8e-89d7-7b439659b9f6,20/07/2022,272.68,Clothing,Debit Card
3013,6a84f1a7-f636-4300-bd8d-c7f888460829,185ee140-348a-4461-bd40-07a7e8535815,2023-10-09,0.00,Health & Beauty,PayPal


In [23]:
social_v2

,CustomerID,InteractionID,InteractionDate,Platform,InteractionType,Sentiment
0,2dcb9523-356b-40b2-a67b-1f27797de261,e5d15761-d0a7-4329-89e3-79a892c56097,2023-07-11,Unknown,Comment,Unknown
1,e12c37b3-7d4d-472f-9fd8-0df2cb3001aa,02f9f376-70ae-4fcd-9070-1db977939948,2023-07-06,Twitter,Share,Unknown
2,08a911a3-65e6-4f5d-a6a1-ae7ddcbe28a2,a83fa04c-f109-4f24-8ce1-2078154f6a1c,2024-05-24,Instagram,Comment,Neutral
3,efdfdfc9-5dbb-4478-911a-101a390a0285,28a69c4b-a2e4-4c74-a130-1132d7733fdf,2023-11-01,Instagram,Like,Neutral
4,ca1e90f6-0e5f-492e-ab92-252ff540da18,d9d1c6f8-5e15-4738-b52b-13c2982420cc,2023-07-08,Instagram,Like,Unknown
...,...,...,...,...,...,...
3015,e06f0e16-8e48-483b-b103-9f105b139ca8,071a2657-12a8-4216-9c64-baf9ab5c9420,2024-04-30,Twitter,Share,Neutral
3016,4d20c2b1-2f8a-4066-902c-bd068c2adccc,e5be1d7a-fdba-4e91-8674-577ac01cddf0,2024-05-16,Unknown,Like,Very Negative
3017,b7b84b81-c61b-4f23-b3db-b18010484122,93d2d3e1-841e-4ece-91c1-3b1998a4cab8,2023-12-16,Twitter,Like,Neutral
3018,5659a8c6-0397-4396-b561-52ec1c8a760b,d7f6806b-f348-431e-9198-07f0685d8d8e,2024-01-28,Facebook,Share,Negative


## Date Validation

In [24]:
demographics_v2["SignupDate"]   = pd.to_datetime(demographics_v2["SignupDate"], errors="coerce")
transactions_v2["TransactionDate"] = pd.to_datetime(transactions_v2["TransactionDate"], errors="coerce")
social_v2["InteractionDate"]   = pd.to_datetime(social_v2["InteractionDate"], errors="coerce")

print("Invalid SignupDate:", demographics_v2["SignupDate"].isna().sum())
print("Invalid TransactionDate:", transactions_v2["TransactionDate"].isna().sum())
print("Invalid InteractionDate:", social_v2["InteractionDate"].isna().sum())

Invalid SignupDate: 98
Invalid TransactionDate: 98
Invalid InteractionDate: 100


In [25]:
before_demo = demographics_v2.shape[0]
before_tx   = transactions_v2.shape[0]
before_sm   = social_v2.shape[0]

demographics_v2 = demographics_v2.dropna(subset=["SignupDate"])
transactions_v2 = transactions_v2.dropna(subset=["TransactionDate"])
social_v2       = social_v2.dropna(subset=["InteractionDate"])

print(f"Removed {before_demo - demographics_v2.shape[0]} rows from demographics")
print(f"Removed {before_tx - transactions_v2.shape[0]} rows from transactions")
print(f"Removed {before_sm - social_v2.shape[0]} rows from social media")

print("All datasets now have valid dates only")

Removed 98 rows from demographics
Removed 98 rows from transactions
Removed 100 rows from social media
All datasets now have valid dates only


## Customer ID's

In [26]:
ids_demo = set(demographics_v2["CustomerID"])
tx_orphans = (~transactions_v2["CustomerID"].isin(ids_demo)).sum()
sm_orphans = (~social_v2["CustomerID"].isin(ids_demo)).sum()

print(f"Transactions without matching demographics: {tx_orphans}")
print(f"Social media interactions without matching demographics: {sm_orphans}")

Transactions without matching demographics: 88
Social media interactions without matching demographics: 92


In [27]:
transactions_before = transactions_v2.shape[0]
social_before = social_v2.shape[0]

transactions_v2 = transactions_v2[transactions_v2["CustomerID"].isin(demographics_v2["CustomerID"])].copy()
social_v2 = social_v2[social_v2["CustomerID"].isin(demographics_v2["CustomerID"])].copy()

print(f"Dropped {transactions_before - transactions_v2.shape[0]} orphan transactions")
print(f"Dropped {social_before - social_v2.shape[0]} orphan social interactions")
print("All remaining records now have matching demographics")

Dropped 88 orphan transactions
Dropped 92 orphan social interactions
All remaining records now have matching demographics


## Standardization

In [28]:
# --- Standardize Categories ---
def clean_str(x):
    if pd.isna(x): return x
    return str(x).strip().title()

In [29]:
# demographics
demographics_v2["IncomeLevel"] = demographics_v2["IncomeLevel"].apply(clean_str)
demographics_v2["Gender"]      = demographics_v2["Gender"].apply(clean_str)

/tmp/ipython-input-770985641.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  demographics_v2["IncomeLevel"] = demographics_v2["IncomeLevel"].apply(clean_str)
/tmp/ipython-input-770985641.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  demographics_v2["Gender"]      = demographics_v2["Gender"].apply(clean_str)


In [30]:
# transactions
transactions_v2["ProductCategory"] = transactions_v2["ProductCategory"].apply(clean_str)
transactions_v2["PaymentMethod"]   = transactions_v2["PaymentMethod"].apply(clean_str)

In [31]:
# social
social_v2["Platform"]  = social_v2["Platform"].apply(clean_str)
social_v2["Sentiment"] = social_v2["Sentiment"].apply(clean_str)

## Outlier

In [32]:
print(transactions_v2["Amount"].describe())

count    2829.000000
mean      492.261730
std       284.707515
min      -100.000000
25%       253.810000
50%       496.625000
75%       725.320000
max       999.860000
Name: Amount, dtype: float64


In [33]:
# Check top 5 largest transactions
print("\nTop 5 transaction amounts:")
print(transactions_v2.nlargest(5, "Amount")[["CustomerID","Amount","ProductCategory"]])


Top 5 transaction amounts:
                                CustomerID  Amount  ProductCategory
745   6f27b068-88f5-4889-b7c4-2c78c9003846  999.86      Electronics
1001  b0afe1c8-686c-4dc8-9668-3b3c33dd364c  999.86       Automotive
2867  9d3466b1-8bcf-430e-a4f9-6a0bad8d296d  999.56  Health & Beauty
1396  a3eb47ea-51aa-4e1a-9d5a-48363ff52433  999.55          Unknown
2008  bd136e0d-1bed-4720-8573-8e62ac1ea02e  998.83         Clothing


In [34]:
amt_threshold = transactions_v2["Amount"].quantile(0.99)
transactions_v2["IsHighSpender"] = (transactions_v2["Amount"] > amt_threshold).astype(int)
print(f"Flagged {transactions_v2['IsHighSpender'].sum()} high-value transactions (>{amt_threshold:.2f})")

Flagged 29 high-value transactions (>991.01)


In [35]:
# Remove negative transactions (likely refunds/errors)
neg_before = transactions_v2.shape[0]
transactions_v2 = transactions_v2[transactions_v2["Amount"] >= 0].copy()
print(f"Dropped {neg_before - transactions_v2.shape[0]} negative-amount transactions")

Dropped 37 negative-amount transactions


In [36]:
# Keep high spenders but flagged
amt_threshold = transactions_v2["Amount"].quantile(0.99)
transactions_v2["IsHighSpender"] = (transactions_v2["Amount"] > amt_threshold).astype(int)
print(f"Flagged {transactions_v2['IsHighSpender'].sum()} high-value transactions (> {amt_threshold:.2f})")

Flagged 28 high-value transactions (> 991.08)


In [37]:
demographics_v2.to_csv("/content/CLEANED_customer_demographics_v3.csv", index=False)
transactions_v2.to_csv("/content/CLEANED_customer_transactions_v3.csv", index=False)
social_v2.to_csv("/content/clean_social_media_interactions_v3.csv", index=False)
print("Saved final v3 datasets")

Saved final v3 datasets


## Load the finalized dataset

In [38]:
demographics_v3 = pd.read_csv("/content/CLEANED_customer_demographics_v3.csv")
transactions_v3 = pd.read_csv("/content/CLEANED_customer_transactions_v3.csv")
social_v3 = pd.read_csv("//content/clean_social_media_interactions_v3.csv")

In [39]:
demographics_v3

,CustomerID,Age,Gender,Location,IncomeLevel,SignupDate
0,9207fa75-5758-48d1-94ad-19c041e0520f,51.0,Female,Jensenberg,Low,2022-11-17
1,5fb09cd8-a473-46f7-80bd-6e49cf509078,45.0,Female,Castilloport,High,2020-07-21
2,c139496e-cc89-498a-bd90-1fb4627b6cff,37.0,Male,Lake Jennifertown,Unknown,2021-01-01
3,50118139-7264-428f-81cc-a25fddc5d6dd,44.0,Male,Port Carl,Medium,2024-06-10
4,7d1f2bbc-8d16-4fbc-9b37-ece3324e8ed4,50.0,Female,Jessebury,High,2023-08-24
...,...,...,...,...,...,...
2920,48313b18-ccdb-4946-b958-1ebfc92dcc2a,59.0,Male,South Andrea,High,2020-02-26
2921,bbe76a26-234f-4a70-b71f-fa9b183f6353,45.0,Male,Gregoryfurt,High,2020-06-23
2922,7eb7f7df-9b0a-416a-829e-bcf4d8bd57dc,55.0,Male,New Anthonyland,High,2022-07-23
2923,ce5cf079-859b-43fa-9609-c6bb7e12c6b7,41.0,Female,Adrianhaven,Low,2024-01-10


In [40]:
print(demographics_v3["Age"].describe())

count    2925.000000
mean       44.749402
std        14.435210
min        18.000000
25%        34.000000
50%        45.000000
75%        57.000000
max        70.000000
Name: Age, dtype: float64


In [41]:
transactions_v3

,CustomerID,TransactionID,TransactionDate,Amount,ProductCategory,PaymentMethod,IsHighSpender
0,60567026-f719-4cd6-849e-137e86d8938f,5ff75116-0a50-4d04-80fb-31e5ccbb0769,2024-05-15,117.640,Clothing,Paypal,0
1,4090ba85-b111-4f75-a792-c777965f5255,2c39b9fe-ff57-4d39-9321-9f5cdf187aa1,2023-04-26,466.140,Health & Beauty,Bank Transfer,0
2,9223891b-73ff-4d5c-b8ae-13ece82ee28b,f79588dd-3db9-4ffa-97f8-7de0e64259f1,2022-09-23,563.990,Clothing,Debit Card,0
3,9243eebc-938f-480c-8564-16d503d250de,401c0fc9-60df-4455-ad78-67c132f9897d,2024-04-15,254.440,Automotive,Paypal,0
4,6e3e8eb8-bc0f-4ffe-9f74-5d5efec9502f,2034aebc-8280-4254-a667-92bcd1c2be4f,2024-06-03,590.520,Home & Garden,Bank Transfer,0
...,...,...,...,...,...,...,...
2787,1ffded5f-b38d-45b7-943d-4100779c3599,6c2c5c2b-7601-41dc-9f56-4009de6f3cea,2023-05-10,591.610,Clothing,Debit Card,0
2788,321acf6c-c401-4f7b-9835-bfb7cd5e48ff,62186b07-9cda-44c5-bf76-bdb0f97aaf39,2023-02-04,496.625,Home & Garden,Bank Transfer,0
2789,f05ec030-ac96-417d-a73c-e3ca3adc1a30,f39c8b1d-595a-4656-8108-25981bf3fa14,2024-03-05,252.700,Electronics,Debit Card,0
2790,086370e7-8884-47d7-89b0-776401a43e1c,652636b6-5c72-44e1-aefb-6adce412ee75,2023-12-05,974.250,Electronics,Debit Card,0


In [42]:
print(transactions_v3["Amount"].describe())

count    2792.000000
mean      500.110471
std       278.246568
min         0.000000
25%       265.537500
50%       496.625000
75%       727.470000
max       999.860000
Name: Amount, dtype: float64


In [43]:
social_v3

,CustomerID,InteractionID,InteractionDate,Platform,InteractionType,Sentiment
0,2dcb9523-356b-40b2-a67b-1f27797de261,e5d15761-d0a7-4329-89e3-79a892c56097,2023-07-11,Unknown,Comment,Unknown
1,e12c37b3-7d4d-472f-9fd8-0df2cb3001aa,02f9f376-70ae-4fcd-9070-1db977939948,2023-07-06,Twitter,Share,Unknown
2,08a911a3-65e6-4f5d-a6a1-ae7ddcbe28a2,a83fa04c-f109-4f24-8ce1-2078154f6a1c,2024-05-24,Instagram,Comment,Neutral
3,efdfdfc9-5dbb-4478-911a-101a390a0285,28a69c4b-a2e4-4c74-a130-1132d7733fdf,2023-11-01,Instagram,Like,Neutral
4,ca1e90f6-0e5f-492e-ab92-252ff540da18,d9d1c6f8-5e15-4738-b52b-13c2982420cc,2023-07-08,Instagram,Like,Unknown
...,...,...,...,...,...,...
2823,e06f0e16-8e48-483b-b103-9f105b139ca8,071a2657-12a8-4216-9c64-baf9ab5c9420,2024-04-30,Twitter,Share,Neutral
2824,4d20c2b1-2f8a-4066-902c-bd068c2adccc,e5be1d7a-fdba-4e91-8674-577ac01cddf0,2024-05-16,Unknown,Like,Very Negative
2825,b7b84b81-c61b-4f23-b3db-b18010484122,93d2d3e1-841e-4ece-91c1-3b1998a4cab8,2023-12-16,Twitter,Like,Neutral
2826,5659a8c6-0397-4396-b561-52ec1c8a760b,d7f6806b-f348-431e-9198-07f0685d8d8e,2024-01-28,Facebook,Share,Negative


In [44]:
print(social_v3["Sentiment"].describe())

count         2828
unique           6
top       Positive
freq           847
Name: Sentiment, dtype: object


## Conclusion

After cleaning the three datasets, we ensured all records are consistent, valid, and analysis-ready. From the demographics file (3,000 rows), we removed 98 entries with invalid signup dates, dropped duplicate customer IDs, and corrected unrealistic ages (e.g., -1, 150) by replacing them with the median, resulting in about 2,902 valid customer records.

In transactions (2,917 rows), 98 rows with invalid dates, 88 orphan records without demographics, and negative amounts were dropped, while 29 high-value transactions (1% above the 99th percentile) were flagged, leaving about 2,730 valid entries.


For social media (3,023 rows), 100 invalid interaction dates and 92 orphan records were removed, with missing platform and sentiment values replaced by “Unknown,” producing about 2,831 valid interactions.

Across all datasets, roughly 6–7% of rows were dropped to ensure integrity, all categorical nulls were standardized, and only realistic numeric values remain, leaving the data consistent and reliable for EDA and machine learning.